controllers/progressivesync_controller.go

```diff
@@ -267,6 +291,35 @@ func (r *ProgressiveSyncReconciler) getClustersFromSelector(ctx context.Context,
 	return secrets, nil
 }
 
+// calculateHashValue returns a hash value string for the ApplicationSet spec
+func (r *ProgressiveSyncReconciler) calculateHashedSpec(ctx context.Context, ps *syncv1alpha1.ProgressiveSync) (string, error) {
+	key := client.ObjectKeyFromObject(ps)
+	latest := syncv1alpha1.ProgressiveSync{}
+	if err := r.Client.Get(ctx, key, &latest); err != nil {
+		r.Log.Error(err, "Failed to retrieve the ProgressiveSync object while calculating the hash value")
+		return "", err
+	}
+
+	appSet := applicationset.ApplicationSet{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: r.ArgoNamespace, Name: latest.Spec.SourceRef.Name}, &appSet); err != nil {
+		r.Log.Error(err, "Failed to retrieve the ApplicationSet object")
+		return "", err
+	}
+
+	appSetSpec := appSet.Spec
+	appSetSpecInBytes, err := json.Marshal(appSetSpec)
+	if err != nil {
+		r.Log.Error(err, "Failed to encode the ApplicanSet spec")
+		return "", err
+	}
+
+	hashedSpecInBytes := md5.Sum(appSetSpecInBytes)
```

Comment:
I don't know if there are any drawbacks about using `md5`. What was the hash function used in kubernetes?

Reply:
I think it boils down to https://github.com/davecgh/go-spew/tree/master/spew, but I struggle to fully understand how they do it. But I don't think it's md5 as there are not references to it. 

About drawbacks, I am not sure. From what I saw, the json is converted to `[]byte` and any changes are visible there. This was enough to convince me that it should be fine and the tests also confirm this. If we want a better understanding, I can spend some more time looking into the `spew` library and try to understand what they do.

Reply:
I spend a bit more time trying to understand it. It seems that it uses FNV-1 and FNV-1a hash functions (https://pkg.go.dev/hash/fnv#New32a). And then spew is the part that expands the fields and writes into it. Or that's effectively that way I see it - it's a bit confusing. (https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_utils.go#L1165)

If we decide that md5 is not good for our purposes, we can just do what we have here - https://github.com/kubernetes/kubernetes/blob/master/pkg/util/hash/hash.go#L28, but this will introduce a dependency on the spew lib.

Reply:
I have been thinking about this, and I guess we can do it without spew. I think they use spew so that they can pass in whole objects and extract the data from them (or at least this is my understanding). Spew also does some formatting, which I am not quite sure about. While in our case, at the moment we work with the json directly in the form of `[]byte` given by the `json.Marshal`. However, I am not 100% sure which approach is better.

Unless there is a disadvantage of using the json, we can do something like - https://play.golang.org/p/T3PIfTeAbTt. Line 16 is the one doing the writing, which in kubernetes is done inside spew. But if that's the case, we can use md5 as well, because I can't really see a lot of difference and benefit on one over the other.

This is my understanding at the moment, but I might be missing or misunderstanding something. I see that you are concerned about the use of `md5`, so maybe we can have a more in-depth discussion to clarify what will be best for us.

PS. Sorry for writing from this acc, I hope it's fine :slightly_smiling_face:

Reply:
No worries, let's go for md5 for the time being.